### PR TITLE
[wip] Python rework

### PR DESCRIPTION
**Pull request by ashtongraves**

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,12 @@
+FROM opensciencegrid/software-base:3.6-el8-release
+RUN yum install -y openssl-devel bzip2-devel libffi-devel wget
+RUN yum groupinstall -y "Development Tools"
+RUN wget https://www.python.org/ftp/python/3.11.4/Python-3.11.4.tgz
```
Comment (djw8605) [inline Dockerfile]: There's gotta be a better way to install python.  Look through `dnf search python` to see if there is a newer version of python.

File: graccbackup@.service [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+Wants=network-online.target
+After=network-online.target
+
+[Service]
+User=gracc
+Group=gracc
+Type=oneshot
+Environment=IN=%i
+PermissionsStartOnly=true
+ExecStartPre=/bin/bash -c "systemctl set-environment DEST=`grep dest /etc/graccarchive/config/gracc-archive-${IN}.toml | grep -v \\# | cut -d\\' -f2`"
```
Comment (djw8605) [inline graccbackup@.service]: I don't think we need this anymore.

File: graccbackup@.service [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+ExecStart=/usr/bin/gracc-backup /var/lib/graccarchive/%i/output /var/lib/graccarchive/%i/secondary ${DEST}
```
Comment (djw8605) [inline graccbackup@.service]: Isn't DESt detected from reading in the TOML file now?

**Review by djw8605 — CHANGES_REQUESTED** [review]
Few comments.

File: gracc-backup.py [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+        os.environ['X509_USER_KEY'] = '/etc/grid-security/backup-cert/gracc.opensciencegrid.org-key.pem'
+        
+        local_full_path = "file://" + file_full_path
+        remote_full_path = output_path + file_name
+        
+        print(subprocess.check_output(f'gfal-copy {local_full_path} {remote_full_path}', shell=True))
+        
+        unformatted_remote_checksum = subprocess.check_output(f'gfal-sum -v {remote_full_path} MD5', shell=True).decode('utf-8')
+        remote_checksum = re.findall('.+ (.+)\\n', unformatted_remote_checksum)[0]
+        unformatted_local_checksum = subprocess.check_output(f'gfal-sum -v {file_full_path} MD5', shell=True).decode('utf-8')
```
Comment (djw8605) [inline gracc-backup.py]: Why are you using gfal-sum to checksum on the localhost?  There are checksum utilities on the command line to do this.

File: graccbackup@.service (hunk 2 above)
Comment (ashtongraves) [reply graccbackup@.service]: FYI Haven't touched the service file yet, this is still the one generated by puppet in use on the gracc server. Was going to do the backup service and dockerfile changes when I'm ready to test it out on the gracc server.

File: gracc-backup.py (hunk 4 above)
Comment (ashtongraves) [reply gracc-backup.py]: 7ccec42d225d5caec4c14ec6a5aa843599710b2d resolves

File: gracc-backup.py
Comment (djw8605) [inline gracc-backup.py:1]: Can you add some comments to this file so we know what's going on?  Maybe split it into functions?

File: Dockerfile [hunk 5]
```diff
@@ -0,0 +1,7 @@
+FROM opensciencegrid/software-base:3.6-el8-release
+RUN dnf install -y python3.11
+ADD test-resources/etc /etc
+ADD test-resources/var /var
+WORKDIR /script
+COPY ./gracc-backup.py .
+ENTRYPOINT ["python3.11", "./gracc-backup.py", "raw"]
```
Comment (djw8605) [inline Dockerfile]: This should probably be just 
ENTRYPOINT ["python3.11", "./gracc-backup.py"]

Then, the "type" of backup will be passed as command line arguments to the docker run command in the .service file.

File: gracc-backup.py [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+
+def backup_archive(file_path, file_name):
+    toml_path = f'/etc/graccarchive/config/gracc-archive-{id}.toml'
+    with open(toml_path, 'rb') as f:
+        toml_text = tomllib.load(f)
+        output_path = toml_text['Directories']['dest']      
+    local_full_path = "file://" + file_path + file_name
+    remote_full_path = output_path + file_name
+    os.environ['X509_USER_CERT'] = '/etc/grid-security/backup-cert/gracc.opensciencegrid.org-cert.pem'
+    os.environ['X509_USER_KEY'] = '/etc/grid-security/backup-cert/gracc.opensciencegrid.org-key.pem'
```
Comment (djw8605) [inline gracc-backup.py]: Check if these environment variables are already set before setting them.  That way we can override the "default" values by setting the environment variable in the docker command.  Probably best to use the `get` command from dictionary.  So something like:

```python
os.environ.get('X509_USER_CERT', '/etc/grid-security/backup-cert/gracc.opensciencegrid.org-cert.pem')
os.environ.get('X509_USER_KEY', '/etc/grid-security/backup-cert/gracc.opensciencegrid.org-key.pem')
```

**Comment by ashtongraves** [comment]
@djw8605 I think this is ready to merge. Has been working

**Pull request merged**